src/transformers/generation/utils.py

```diff
@@ -1322,6 +1322,14 @@ def generate(
         )
         batch_size = inputs_tensor.shape[0]
 
+        ALL_CACHE_CLASSES = {
+            "static": StaticCache
+        }
+        if generation_config.cache_implementation in ALL_CACHE_CLASSES and not model_kwargs.get("past_key_values", False):
+            cache_cls = ALL_CACHE_CLASSES[generation_config.cache_implementation]
+            self._setup_cache(cache_cls, batch_size)
```

Comment:
I guess this is the big design change which instantiates a `cache_cls` to each attn class. I understand why this is necessary and think it's ok to do this for the static cache, but I think we should not do it for the dynamic or other caches because it'll prevent the model from being used in a multi-threading environment. We can simply say that when enabling the static cache, multi-threading is not possible, but we should probably keep the dynamic cache the way it is so that the model can be used in a multi-threading environment

Reply:
IIRC, only if `generation_config.cache_implementation == "static"` will have `model_kwargs["past_key_values"] = cache_cls(self.config, max_batch_size=batch_size)`. So I'm not seeing the problem, unless @ArthurZucker  plan to extend `ALL_CACHE_CLASSES` in the future.

BTW, what's the issue with the  multi-threading when a `cache_cls` is used?